### PR TITLE
Add a test to check max_rows_to_read_leaf behaviour

### PR DESCRIPTION
**Pull request by seandhaynes**
In ClickHouse v22.3, querying a distributed table through a view results in a max_rows_to_read_leaf limit reached exception being raised on the root executor despite the row limit for each leaf node not being reached.

The problem looks to be been a bug in the SourceWIthProcess processor that was used when selecting via a view, and checked the total number of rows read against the leaf limit setting: https://github.com/ClickHouse/ClickHouse/blob/22.3/src/Processors/Sources/SourceWithProgress.cpp#L108,L126

Which looks to be fixed in v22.8 onwards in this PR that moves this logic into the Pipeline executor: https://github.com/ClickHouse/ClickHouse/pull/37543/

An example using any v22.3 ClickHouse build:
```
CREATE TABLE local_table ON CLUSTER `{cluster}`
(
    `id` UInt32,
    `d` DateTime
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/0/max_rows_read_leaf', '{replica}')
PARTITION BY toYYYYMM(d)
ORDER BY d

Query id: 97d6e58f-a98c-48a9-b066-a2f9de30326a

┌─host──────────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ clickhouse-06 │ 9000 │      0 │       │                   5 │                0 │
│ clickhouse-03 │ 9000 │      0 │       │                   4 │                0 │
│ clickhouse-02 │ 9000 │      0 │       │                   3 │                0 │
│ clickhouse-05 │ 9000 │      0 │       │                   2 │                0 │
│ clickhouse-04 │ 9000 │      0 │       │                   1 │                0 │
│ clickhouse-01 │ 9000 │      0 │       │                   0 │                0 │
└───────────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

6 rows in set. Elapsed: 0.304 sec.
```

```
CREATE TABLE distributed_table ON CLUSTER `{cluster}`
(
    `id` UInt32,
    `d` DateTime
)
ENGINE = Distributed('{cluster}', default, local_table)

Query id: 41790cab-d877-47bc-ba56-558684db7a6e

┌─host──────────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ clickhouse-06 │ 9000 │      0 │       │                   5 │                0 │
│ clickhouse-02 │ 9000 │      0 │       │                   4 │                0 │
│ clickhouse-03 │ 9000 │      0 │       │                   3 │                0 │
│ clickhouse-05 │ 9000 │      0 │       │                   2 │                0 │
│ clickhouse-04 │ 9000 │      0 │       │                   1 │                0 │
│ clickhouse-01 │ 9000 │      0 │       │                   0 │                0 │
└───────────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

6 rows in set. Elapsed: 0.144 sec.
```

```
CREATE OR REPLACE VIEW test_view ON CLUSTER `{cluster}` AS
SELECT id
FROM distributed_table

Query id: c585b39a-6284-4bf0-9b5e-6a61f660b387

┌─host──────────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ clickhouse-06 │ 9000 │      0 │       │                   5 │                0 │
│ clickhouse-02 │ 9000 │      0 │       │                   4 │                0 │
│ clickhouse-03 │ 9000 │      0 │       │                   3 │                0 │
│ clickhouse-05 │ 9000 │      0 │       │                   2 │                0 │
│ clickhouse-04 │ 9000 │      0 │       │                   1 │                0 │
│ clickhouse-01 │ 9000 │      0 │       │                   0 │                0 │
└───────────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

6 rows in set. Elapsed: 0.164 sec.
```

INSERT 400 rows across 2 shards:
```
clickhouse-01 :) INSERT INTO local_table (id) select * from system.numbers limit 200

INSERT INTO local_table (id) SELECT *
FROM system.numbers
LIMIT 200

Query id: 153f9a6b-0955-47c6-84d5-64da1c468f84

Ok.

0 rows in set. Elapsed: 0.040 sec.

clickhouse-04 :) INSERT INTO local_table (id) select * from system.numbers limit 200

INSERT INTO local_table (id) SELECT *
FROM system.numbers
LIMIT 200

Query id: 3975ac5a-0da9-42ac-be32-e9fbe72f15a3

Ok.
```

Queries using the distributed table (succeeds as expected):
```
clickhouse-01 :) select count() from distributed_table SETTINGS max_rows_to_read_leaf=200;

SELECT count()
FROM distributed_table
SETTINGS max_rows_to_read_leaf = 200

Query id: 5bfe13a6-d70b-4705-899a-bcafb7c8fb04

┌─count()─┐
│     400 │
└─────────┘

1 rows in set. Elapsed: 0.110 sec.
```
Query to the view that selects from the distributed table fails with a limit error when it should not:
```
clickhouse-01 :) SELECT count() from test_view SETTINGS max_rows_to_read_leaf=200

SELECT count()
FROM test_view
SETTINGS max_rows_to_read_leaf = 200

Query id: 2229243f-46d4-4fd5-80a0-636b34dc12fd


0 rows in set. Elapsed: 0.061 sec.

Received exception from server (version 22.3.8):
Code: 158. DB::Exception: Received from localhost:9000. DB::Exception: Limit for rows or bytes to read on leaf node exceeded, max rows: 200.00, current rows: 400.00: While executing Remote. (TOO_MANY_ROWS)
```

**This is fixed in ClickHouse v23.3, but Adding this integration test to catch any issues in the future.**

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
Add a test to check max_rows_to_read_leaf behaviour

### CODE REVIEW

**Comment by seandhaynes** [comment]
Associated issue: https://github.com/ClickHouse/ClickHouse/issues/48965

File: tests/integration/test.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from contextlib import contextmanager
```
Comment (hanfei1991) [inline tests/integration/test.py:1]: should this file be `tests/integration/test_max_rows_to_read_leaf_with_view/test.py`?

Comment (seandhaynes) [reply tests/integration/test.py]: > should this file be `tests/integration/test_max_rows_to_read_leaf_with_view/test.py`?

🤦  Yes, it should. Sorry about that @hanfei1991 . Updated my branch.

**Comment by seandhaynes** [comment]
@hanfei1991 Thank for you adding the "can be tested" flag. Sorry I missed that Black was used for formatting. I have black-formatted my test and re-pushed.

**Comment by seandhaynes** [comment]
@hanfei1991 Hey, I seem to have several build failures that don't look related to my change. For example, an S3 merge tree test that fails on a cleanup test: https://s3.amazonaws.com/clickhouse-test-reports/48950/f861d7b587495e85501bb11edc9ee2b5e0e7834c/integration_tests__release__[2/4]/integration_run_test_merge_tree_s3_test_py_0.log

Can these be retried, or is there anymore I can do?

**Comment by seandhaynes** [comment]
@hanfei1991 sorry to trouble you again, but looks like there is only a single test failure now, and that test is currently expected to fail. Are you happy for this to be merged, or is there anything I can do to help with this last remaining test?

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 1d51bc90f0e7e502477508b20254c559c54946a2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48950/1d51bc90f0e7e502477508b20254c559c54946a2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by hanfei1991 — APPROVED** [review]

**Comment by seandhaynes** [comment]
Thanks very much for your help, @hanfei1991 !

**Pull request merged**